### PR TITLE
Feature to update all pull requests asynchronously

### PR DESCRIPTION
**Pull request by Tantalon**
Thanks for this project! We want to rebase PRs automatically and although there's lots of GitHub actions to do so, this was the only one I reviewed that uses the GitHub API.

This pull request adds two features we need:
* Update all open pull requests if number is not specified
* A flag to run asynchronously rather than waiting

### CODE REVIEW

File: action.yml [hunk 1]
```diff
@@ -10,7 +10,11 @@ inputs:
     required: true
   number:
     description: 'Pull Request number'
-    required: true
+    required: false
+  async:
```
Comment (juliangruber) [inline action.yml]: I think a better name would be `waitForPullRequestUpdated`, which defaults to `true`. Iiuc this doesn't really have anything to do with synchronous vs asynchronous

File: README.md [hunk 2]
```diff
@@ -1,12 +1,13 @@
 # update-pull-request-branch-action
 
-Update a Pull Request from its base branch.
+Update pull requests when their base branches change.
 
-After updating it waits until the branch has been successfully updated,
+Optionally, waits until the branch has been successfully updated,
```
Comment (juliangruber) [inline README.md]:
```suggestion
Optionally waits until the branch has been successfully updated,
```

File: README.md [hunk 3]
```diff
@@ -1,12 +1,13 @@
 # update-pull-request-branch-action
 
-Update a Pull Request from its base branch.
+Update pull requests when their base branches change.
```
Comment (juliangruber) [inline README.md]: This isn't true, updating when the base branch changes is your specific use case, but this action is more generic than that

File: index.js [hunk 4]
```diff
@@ -6,32 +6,59 @@ const { GitHub, context } = require('@actions/github')
 const sleep = dt => new Promise(resolve => setTimeout(resolve, dt))
 
 const main = async () => {
-  const token = core.getInput('github-token')
+  // GitHub token, also check underscore notation for console testing
+  const token = core.getInput('github-token') || core.getInput('github_token')
```
Comment (juliangruber) [inline index.js]: Can you please explain when `github_token` would be supplied instead of `github-token`?

**Review by juliangruber — CHANGES_REQUESTED** [review]

File: index.js (hunk 4 above)
Comment (Tantalon) [reply index.js]: Testing it from the console like this needs the variable name to not contain a hyphen:
INPUT_GITHUB_TOKEN=... GITHUB_REPOSITORY=Tantalon/spike-github-autorebase node index.js

File: README.md (hunk 3 above)
Comment (Tantalon) [reply README.md]: Agreed, I've reverted the sentence.

File: action.yml (hunk 1 above)
Comment (Tantalon) [reply action.yml]: Updated to waitForPullRequestUpdated. I'll retest then resolve this comment.

File: index.js (hunk 4 above)
Comment (juliangruber) [reply index.js]: Ah, I see. I've never run this in my local console. Could we just rename the input then (so that it works for dev and prod), and make this a major release?

Comment (Tantalon) [reply index.js]: Absolutely, I've committed this change.

File: index.js [hunk 5]
```diff
@@ -6,32 +6,59 @@ const { GitHub, context } = require('@actions/github')
...
       ...context.repo,
       pull_number: number
     })
-    if (res.data.head.sha !== oldSha) return
+    const oldSha = res.data.head.sha
+    const prMergeable = res.data.mergeable
+    const prMergeableState = res.data.mergeable_state
+
+    // merge automatically if pull request is outdated and checks have passed
+    if (prMergeable && prMergeableState === 'behind') {
```
Comment (juliangruber) [inline index.js:37]: this is also out of scope, there are different actions for auto merging PRs

**Review by juliangruber — CHANGES_REQUESTED** [review]

File: index.js (hunk 5 above)
Comment (Tantalon) [reply index.js:37]: mergeable_state tells us whether or not a pull request can be updated from its base branch

clean = no update is needed
behind = it's out-of-date and can be updated
dirty = it's out-of-date but can't be updated because of a conflict
unknown = GitHub is still checking it

from here: https://github.com/octokit/octokit.net/issues/1763

Comment (Tantalon) [reply index.js:37]: Oh, but I realise the comment I put in the code is misleading and wrong, it should say update not merge

Comment (juliangruber) [reply index.js:37]: I understand, but I mean there are different github actions for merging a PR once it's green and up to date with the base branch. Let's not do more than we need here in this action, its purpose is to solely update branches

Comment (Tantalon) [reply index.js:37]: Makes sense. I've updated the comment to reflect that we're just checking whether the pull request can be updated. Are you happy with this or is more needed to resolve this part?

File: index.js [hunk 6]
```diff
@@ -6,32 +6,59 @@ const { GitHub, context } = require('@actions/github')
...
+          core.debug('sleep')
+          await sleep(1000)
 
-    core.debug('sleep')
-    await sleep(1000)
+          const res = await octokit.pulls.get({
+            ...context.repo,
+            pull_number: number
+          })
+          prUpdated = res.data.head.sha !== oldSha
```
Comment (juliangruber) [inline index.js]: instead of using state, can simply use `break` to exit the loop

**Review by juliangruber — CHANGES_REQUESTED** [review]

File: index.js (hunk 6 above)
Comment (Tantalon) [reply index.js]: I've updated it

File: README.md [hunk 7]
```diff
@@ -2,18 +2,29 @@
...
 steps:
   - name: Update Pull Request
     uses: juliangruber/update-pull-request-branch-action@v1
     with:
       number: 1
-      github-token: ${{ secrets.GITHUB_TOKEN }}
+      github_token: ${{ secrets.GITHUB_TOKEN }}
+```
+
+To update all pull requests without waiting:
```
Comment (juliangruber) [inline README.md]:
```suggestion
To update all open pull requests without waiting:
```

**Review by juliangruber — CHANGES_REQUESTED** [review]
Just a tiny last change, then I'll publish 🚀

**Comment by juliangruber** [comment]
Thanks for the great work! https://github.com/juliangruber/update-pull-request-branch-action/releases/tag/v1.1.0
